lib/model/query/entities.js

```diff
@@ -89,9 +89,53 @@ const createSource = (details = null, subDefId = null, eventId = null) => ({ one
     .then((row) => row.id);
 };
 
+//
+// Entity event processing
+//
+
+const _createEntity = (dataset, entityData, submissionId, submissionDef, submissionDefId, event, parentEvent) => async ({ Entities, Audits }) => {
+  const partial = await Entity.fromParseEntityData(entityData);
+
+  const sourceDetails = { submission: { instanceId: submissionDef.instanceId }, parentEventId: parentEvent ? parentEvent.id : undefined };
+  const sourceId = await Entities.createSource(sourceDetails, submissionDefId, event.id);
+  const entity = await Entities.createNew(dataset, partial, submissionDef, sourceId);
+
+  return Audits.log({ id: event.actorId }, 'entity.create', { acteeId: dataset.acteeId },
+    {
+      entityId: entity.id, // Added in v2023.3 and backfilled
+      entityDefId: entity.aux.currentVersion.id, // Added in v2023.3 and backfilled
+      entity: { uuid: entity.uuid, dataset: dataset.name },
+      submissionId,
+      submissionDefId
+    });
+};
+
+const _updateEntity = (dataset, entityData, submissionId, submissionDef, submissionDefId, event, parentEvent) => async ({ Entities }) => {
+
+  // Get version of entity on the server
+  const serverEntity = await Entities.getById(dataset.id, entityData.system.id).then(o => o.get());
+
+  // merge data
+  const mergedData = mergeRight(serverEntity.aux.currentVersion.data, entityData.data);
+  const mergedLabel = entityData.system.label || serverEntity.aux.currentVersion.label;
```

Comment:
The updates should be applied on top of the specified base version, not on top of the current version. But we could come back to that.

Reply:
that sounds like it would go against "simultaneous updates"